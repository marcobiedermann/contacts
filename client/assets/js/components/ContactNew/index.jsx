import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';

class ContactNew extends Component {
  constructor(props) {
    super(props);

    this.state = {};

    this.onSubmit = this.onSubmit.bind(this);
    this.onInput = this.onInput.bind(this);
  }

  onSubmit(event) {
    event.preventDefault();
    const {
      addContact,
      history,
    } = this.props;
    const {
      firstName,
      lastName,
      streetAddress,
      zipCode,
      city,
      country,
      email,
      phone,
    } = this.state;

    const contact = {
      name: {
        firstName,
        lastName,
      },
      address: {
        streetAddress,
        zipCode,
        city,
        country,
      },
      email,
      phone,
    };

    addContact(contact);
    history.push('/contacts');
  }

  onInput(event) {
    const {
      name,
      value,
    } = event.target;

    this.setState({
      [name]: value,
    });
  }

  render() {
    return (
      <form onSubmit={this.onSubmit}>
        <div>
          <label htmlFor="firstName">
            Firstname
            <input
              type="text"
              id="firstName"
              name="firstName"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="lastName">
            Lastname
            <input
              type="text"
              id="lastName"
              name="lastName"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="streetAddress">
            Street
            <input
              type="text"
              id="streetAddress"
              name="streetAddress"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="zipCode">
            Zip
            <input
              type="text"
              id="zipCode"
              name="zipCode"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="city">
            City
            <input
              type="text"
              id="city"
              name="city"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="country">
            Country
            <input
              type="text"
              id="country"
              name="country"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="email">
            Email
            <input
              type="email"
              id="email"
              name="email"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <label htmlFor="phone">
            Phone
            <input
              type="tel"
              id="phone"
              name="phone"
              onInput={this.onInput}
            />
          </label>
        </div>

        <div>
          <input
            type="submit"
            value="Save new contact"
          />
        </div>
      </form>
    );
  }
}

ContactNew.propTypes = {
  addContact: PropTypes.func,
  history: PropTypes.shape(),
};

ContactNew.defaultProps = {
  addContact: () => {},
  history: null,
};

export default withRouter(ContactNew);
