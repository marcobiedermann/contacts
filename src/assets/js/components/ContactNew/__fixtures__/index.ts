import { ContactNewProps } from '..';
import validationSchema from '../../../utils/validation';

const props: ContactNewProps = {
  initialValues: {
    addresses: [
      {
        type: '',
        value: {
          city: '',
          country: '',
          street: '',
          zip: '',
        },
      },
    ],
    emails: [
      {
        type: '',
        value: '',
      },
    ],
    name: {
      firstName: '',
      lastName: '',
    },
    phones: [
      {
        type: '',
        value: '',
      },
    ],
  },
  onSubmit: () => ({}),
  validationSchema,
};

export default props;
