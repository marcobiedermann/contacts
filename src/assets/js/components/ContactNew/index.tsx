/* eslint-disable react/no-array-index-key */

import { Field, FieldArray, Form, Formik, FormikHelpers } from 'formik';
import React, { FC } from 'react';
import { useTranslation } from 'react-i18next';
import { AddressProps } from '../Address';
import { EmailProps } from '../Email';
import Label from '../Label';
import { PhoneProps } from '../Phone';
import styles from './style.module.css';

export interface Values {
  addresses: AddressProps[];
  emails: EmailProps[];
  name: {
    firstName: string;
    lastName: string;
  };
  phones: PhoneProps[];
}

export interface ContactNewProps {
  initialValues: Values;
  onSubmit: (values: Values, formikHelpers: FormikHelpers<Values>) => void;
}

const ContactNew: FC<ContactNewProps> = (props) => {
  const { initialValues, onSubmit } = props;
  const { t } = useTranslation();

  return (
    <Formik initialValues={initialValues} onSubmit={onSubmit}>
      {(formikProps) => {
        const { isSubmitting, values } = formikProps;

        return (
          <Form className={styles.form}>
            <div className={styles.form__field}>
              <Label htmlFor="name.firstName">Firstname</Label>
              <Field className={styles.form__input} id="name.firstName" name="name.firstName" type="text" />
            </div>

            <div className={styles.form__field}>
              <Label htmlFor="name.lastName">Lastname</Label>
              <Field className={styles.form__input} id="name.lastName" name="name.lastName" type="text" />
            </div>

            <FieldArray name="addresses">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <div>
                    {values.addresses.map((_address, index) => {
                      const id = `addresses.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.street`}>Street</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.street`}
                              name={`${id}.value.street`}
                              type="text"
                            />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.zip`}>Zip</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.zip`}
                              name={`${id}.value.zip`}
                              type="text"
                            />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.city`}>City</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.city`}
                              name={`${id}.value.city`}
                              type="text"
                            />
                          </div>

                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value.country`}>Country</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value.country`}
                              name={`${id}.value.country`}
                              type="text"
                            />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: {
                            city: '',
                            country: '',
                            street: '',
                            zip: '',
                          },
                        })
                      }
                    >
                      +
                    </button>
                  </div>
                );
              }}
            </FieldArray>

            <FieldArray name="emails">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <div>
                    {values.emails.map((_email, index) => {
                      const id = `emails.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value`}>Email</Label>
                            <Field
                              className={styles.form__input}
                              id={`${id}.value`}
                              name={`${id}.value`}
                              type="email"
                            />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: '',
                        })
                      }
                    >
                      +
                    </button>
                  </div>
                );
              }}
            </FieldArray>

            <FieldArray name="phones">
              {(arrayHelpers) => {
                const { push, remove } = arrayHelpers;

                return (
                  <div>
                    {values.phones.map((_phone, index) => {
                      const id = `phones.${index}`;

                      return (
                        <div key={id}>
                          <div className={styles.form__field}>
                            <Label htmlFor={`${id}.value`}>Phone</Label>
                            <Field className={styles.form__input} id={`${id}.value`} name={`${id}.value`} type="tel" />
                          </div>

                          <button type="button" onClick={() => remove(index)}>
                            -
                          </button>
                        </div>
                      );
                    })}

                    <button
                      type="button"
                      onClick={() =>
                        push({
                          type: '',
                          value: '',
                        })
                      }
                    >
                      +
                    </button>
                  </div>
                );
              }}
            </FieldArray>

            <div className={styles.form__field}>
              <button disabled={isSubmitting} type="submit">
                {t('common:save')}
              </button>
            </div>
          </Form>
        );
      }}
    </Formik>
  );
};

export default ContactNew;
