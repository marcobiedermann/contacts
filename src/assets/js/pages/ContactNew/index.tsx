import { FormikHelpers } from 'formik';
import React, { useCallback } from 'react';
import { Helmet } from 'react-helmet';
import { useTranslation } from 'react-i18next';
import { useFirestore } from 'reactfire';
import ContactNew, { Values } from '../../components/ContactNew';
import validationSchema from '../../utils/validation';

function ContactNewPage(): JSX.Element {
  const firestore = useFirestore();
  const contactsCollection = firestore.collection('contacts');
  const { t } = useTranslation();

  const initialValues: Values = {
    addresses: [
      {
        type: '',
        value: {
          city: '',
          country: '',
          street: '',
          zip: '',
        },
      },
    ],
    emails: [
      {
        type: '',
        value: '',
      },
    ],
    name: {
      firstName: '',
      lastName: '',
    },
    phones: [
      {
        type: '',
        value: '',
      },
    ],
  };

  const onSubmit = useCallback((values: Values, formikHelpers: FormikHelpers<Values>) => {
    const { setSubmitting } = formikHelpers;

    contactsCollection.add(values);

    setSubmitting(false);
  }, []);

  return (
    <>
      <Helmet>
        <title>{t('pages:new')}</title>
      </Helmet>
      <ContactNew initialValues={initialValues} onSubmit={onSubmit} validationSchema={validationSchema} />
    </>
  );
};

export default ContactNewPage;
