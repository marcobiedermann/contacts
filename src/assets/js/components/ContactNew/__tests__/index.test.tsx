import { render } from '@testing-library/react';
import React from 'react';
import ContactNew from '..';
import props from '../__fixtures__';

describe('contact new component', () => {
  it('should render correctly', () => {
    expect.assertions(1);

    const { container } = render(<ContactNew {...props} />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
