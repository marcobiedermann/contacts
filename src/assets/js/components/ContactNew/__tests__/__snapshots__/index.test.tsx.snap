// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`contact new component should render correctly 1`] = `
<form
  action="#"
  class="form"
>
  <fieldset>
    <legend>
      Name
    </legend>
    <div
      class="form__field"
    >
      <label
        class="label"
        for="name.firstName"
      >
        Firstname
      </label>
      <input
        class="form__input"
        id="name.firstName"
        name="name.firstName"
        type="text"
        value=""
      />
    </div>
    <div
      class="form__field"
    >
      <label
        class="label"
        for="name.lastName"
      >
        Lastname
      </label>
      <input
        class="form__input"
        id="name.lastName"
        name="name.lastName"
        type="text"
        value=""
      />
    </div>
  </fieldset>
  <fieldset>
    <legend>
      Addresses
    </legend>
    <div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="addresses.0.type"
        >
          Type
        </label>
        <select
          class="form__input"
          id="addresses.0.type"
          name="addresses.0.type"
        >
          <option
            value="home"
          >
            Home
          </option>
          <option
            value="work"
          >
            Work
          </option>
        </select>
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="addresses.0.value.street"
        >
          Street
        </label>
        <input
          class="form__input"
          id="addresses.0.value.street"
          name="addresses.0.value.street"
          type="text"
          value=""
        />
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="addresses.0.value.zip"
        >
          Zip
        </label>
        <input
          class="form__input"
          id="addresses.0.value.zip"
          name="addresses.0.value.zip"
          type="text"
          value=""
        />
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="addresses.0.value.city"
        >
          City
        </label>
        <input
          class="form__input"
          id="addresses.0.value.city"
          name="addresses.0.value.city"
          type="text"
          value=""
        />
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="addresses.0.value.country"
        >
          Country
        </label>
        <input
          class="form__input"
          id="addresses.0.value.country"
          name="addresses.0.value.country"
          type="text"
          value=""
        />
      </div>
      <button
        type="button"
      >
        -
      </button>
    </div>
    <button
      type="button"
    >
      +
    </button>
  </fieldset>
  <fieldset>
    <legend>
      Emails
    </legend>
    <div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="emails.0.type"
        >
          Type
        </label>
        <select
          class="form__input"
          id="emails.0.type"
          name="emails.0.type"
        >
          <option
            value="home"
          >
            Home
          </option>
          <option
            value="work"
          >
            Work
          </option>
        </select>
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="emails.0.value"
        >
          Email
        </label>
        <input
          class="form__input"
          id="emails.0.value"
          name="emails.0.value"
          type="email"
          value=""
        />
      </div>
      <button
        type="button"
      >
        -
      </button>
    </div>
    <button
      type="button"
    >
      +
    </button>
  </fieldset>
  <fieldset>
    <legend>
      Phones
    </legend>
    <div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="phones.0.type"
        >
          Type
        </label>
        <select
          class="form__input"
          id="phones.0.type"
          name="phones.0.type"
        >
          <option
            value="home"
          >
            Home
          </option>
          <option
            value="work"
          >
            Work
          </option>
        </select>
      </div>
      <div
        class="form__field"
      >
        <label
          class="label"
          for="phones.0.value"
        >
          Phone
        </label>
        <input
          class="form__input"
          id="phones.0.value"
          name="phones.0.value"
          type="tel"
          value=""
        />
      </div>
      <button
        type="button"
      >
        -
      </button>
    </div>
    <button
      type="button"
    >
      +
    </button>
  </fieldset>
  <div
    class="form__field"
  >
    <button
      type="submit"
    >
      common:save
    </button>
  </div>
</form>
`;
